File: back/src/main/java/com/triptrace/domain/marker/marker/controller/ApiV1MarkerController.java
package com.triptrace.domain.marker.marker.controller;

import com.triptrace.domain.marker.marker.dto.MarkerCreateRequest;
import com.triptrace.domain.marker.marker.dto.MarkerModifyRequest;
import com.triptrace.domain.marker.marker.dto.MarkerResponse;
import com.triptrace.domain.marker.marker.dto.PlaceCandidateResponse;
import com.triptrace.domain.marker.marker.service.MarkerService;
import com.triptrace.global.rsData.RsData;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1")
public class ApiV1MarkerController {

    private final MarkerService markerService;

    // 마커 생성
    @PostMapping("/posts/{postId}/markers")
    public RsData<MarkerResponse> createMarker(
        @PathVariable Long postId,
        @AuthenticationPrincipal Long memberId,
        @Valid @RequestBody MarkerCreateRequest request
    ) {

        MarkerResponse response = markerService.createMarker(postId, memberId, request);

        return new RsData<>(
            "201-04",
            "마커가 생성되었습니다.",
            response
        );
    }

    // 마커 목록 조회
    @GetMapping("/posts/{postId}/markers")
    public RsData<List<MarkerResponse>> getMarkers (
        @PathVariable Long postId
    ) {

        return new RsData<>(
            "200-04",
            "마커 목록 조회에 성공했습니다.",
            markerService.getMarkers(postId)
        );
    }

    // 마커 상세 조회
    @GetMapping("/posts/markers/{markerId}")
    public RsData<MarkerResponse> getMarker (
        @PathVariable Long markerId
    ) {

        return new RsData<>(
            "200-04",
            "마커 조회에 성공했습니다.",
            markerService.getMarker(markerId)
        );
    }

    // 마커 장소명 후보 조회
    @GetMapping("/posts/markers/{markerId}/place-candidates")
    public RsData<List<PlaceCandidateResponse>> getPlaceCandidates(
        @PathVariable Long markerId,
        @AuthenticationPrincipal Long memberId
    ) {

        return new RsData<>(
            "200-04",
            "마커 장소명 후보 조회에 성공했습니다.",
            markerService.getPlaceCandidates(markerId, memberId)
        );
    }

    // 장소 검색
    @GetMapping("/places/search")
    public RsData<List<PlaceCandidateResponse>> searchPlaces(
        @RequestParam String keyword
    ) {

        return new RsData<>(
            "200-04",
            "장소 검색에 성공했습니다.",
            markerService.searchPlaces(keyword)
        );
    }

    // 좌표 기준 주변 장소 조회
    @GetMapping("/places/nearby")
    public RsData<List<PlaceCandidateResponse>> findNearbyPlaces(
        @RequestParam BigDecimal latitude,
        @RequestParam BigDecimal longitude
    ) {

        return new RsData<>(
            "200-04",
            "주변 장소 조회에 성공했습니다.",
            markerService.findNearbyPlaces(latitude, longitude)
        );
    }

    // 마커 수정
    @PatchMapping("/posts/markers/{markerId}")
    public RsData<MarkerResponse> modifyMarker (
        @PathVariable Long markerId,
        @AuthenticationPrincipal Long memberId,
        @Valid @RequestBody MarkerModifyRequest request
    ) {

        return new RsData<>(
            "200-04",
            "마커가 수정되었습니다.",
            markerService.modifyMarker(markerId, memberId, request)
        );
    }

    @DeleteMapping("/posts/markers/{markerId}")
    public void deleteMarker(
        @PathVariable Long markerId,
        @AuthenticationPrincipal Long memberId
    ) {
        markerService.deleteMarker(markerId, memberId);
    }
}

File: back/src/main/kotlin/com/triptrace/domain/marker/marker/controller/ApiV1MarkerController.kt
package com.triptrace.domain.marker.marker.controller

import com.triptrace.domain.marker.marker.dto.MarkerCreateRequest
import com.triptrace.domain.marker.marker.dto.MarkerModifyRequest
import com.triptrace.domain.marker.marker.dto.MarkerResponse
import com.triptrace.domain.marker.marker.dto.PlaceCandidateResponse
import com.triptrace.domain.marker.marker.service.MarkerService
import com.triptrace.global.rsData.RsData
import jakarta.validation.Valid
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.math.BigDecimal

@RestController
@RequestMapping("/api/v1")
class ApiV1MarkerController(
    private val markerService: MarkerService,
) {
    @PostMapping("/posts/{postId}/markers")
    fun createMarker(
        @PathVariable postId: Long,
        @AuthenticationPrincipal memberId: Long,
        @Valid @RequestBody request: MarkerCreateRequest,
    ): RsData<MarkerResponse> = RsData("201-04", "마커가 생성되었습니다.", markerService.createMarker(postId, memberId, request))

    @GetMapping("/posts/{postId}/markers")
    fun getMarkers(
        @PathVariable postId: Long,
    ): RsData<List<MarkerResponse>> = RsData("200-04", "마커 목록 조회에 성공했습니다.", markerService.getMarkers(postId))

    @GetMapping("/posts/markers/{markerId}")
    fun getMarker(
        @PathVariable markerId: Long,
    ): RsData<MarkerResponse> = RsData("200-04", "마커 조회에 성공했습니다.", markerService.getMarker(markerId))

    @GetMapping("/posts/markers/{markerId}/place-candidates")
    fun getPlaceCandidates(
        @PathVariable markerId: Long,
        @AuthenticationPrincipal memberId: Long,
    ): RsData<List<PlaceCandidateResponse>> =
        RsData("200-04", "마커 장소명 후보 조회에 성공했습니다.", markerService.getPlaceCandidates(markerId, memberId))

    @GetMapping("/places/search")
    fun searchPlaces(
        @RequestParam keyword: String,
    ): RsData<List<PlaceCandidateResponse>> = RsData("200-04", "장소 검색에 성공했습니다.", markerService.searchPlaces(keyword))

    @GetMapping("/places/nearby")
    fun findNearbyPlaces(
        @RequestParam latitude: BigDecimal,
        @RequestParam longitude: BigDecimal,
    ): RsData<List<PlaceCandidateResponse>> = RsData("200-04", "주변 장소 조회에 성공했습니다.", markerService.findNearbyPlaces(latitude, longitude))

    @PatchMapping("/posts/markers/{markerId}")
    fun modifyMarker(
        @PathVariable markerId: Long,
        @AuthenticationPrincipal memberId: Long,
        @Valid @RequestBody request: MarkerModifyRequest,
    ): RsData<MarkerResponse> = RsData("200-04", "마커가 수정되었습니다.", markerService.modifyMarker(markerId, memberId, request))

    @DeleteMapping("/posts/markers/{markerId}")
    fun deleteMarker(
        @PathVariable markerId: Long,
        @AuthenticationPrincipal memberId: Long,
    ) = markerService.deleteMarker(markerId, memberId)
}
